pkg/agent/controller/service_import.go

```diff
@@ -267,18 +271,24 @@ func (c *ServiceImportController) reconcileLocalAggregatedServiceImports() {
 	})
 }
 
-func (c *ServiceImportController) startEndpointsController(serviceImport *mcsv1a1.ServiceImport) error {
+func (c *ServiceImportController) startEndpointsController(ctx context.Context, serviceImport *mcsv1a1.ServiceImport) error {
 	key, _ := cache.MetaNamespaceKeyFunc(serviceImport)
 
-	if obj, found := c.endpointControllers.LoadAndDelete(key); found {
-		logger.V(log.DEBUG).Infof("Stopping previous endpoints controller for %q", key)
-		obj.(*ServiceEndpointSliceController).stop()
+	if obj, found := c.endpointControllers.Load(key); found {
+		logger.V(log.DEBUG).Infof("Stopping previous EndpointSlice controller for %q", key)
+
+		err := obj.(*ServiceEndpointSliceController).stop(ctx)
+		if err != nil {
+			return errors.Wrapf(err, "failed to stop previous EndpointSlice controller for %q", key)
```

Comment:
In what scenarios this can fail? Is it safe to delete if failed to stop previous? Any cleanup missed?

Reply:
> In what scenarios this can fail?

If `stop` times out while awaiting the syncer queue to drain. It may mean it's stuck trying to create an EPS. We propagate the error so it is retried and give the stuck process a chance to complete, successfully or not.

> Is it safe to delete if failed to stop previous?

I'm not clear - safe to delete what?

> Any cleanup missed?

There's no more cleanup here.